### DISCUSSION TITLE
Oracle ORM JSON in CLOB

### DISCUSSION THREAD

**Discussion by Oynowan** [sqlalchemy/sqlalchemy#9112] [Usage Questions]
I have an Oracle database with json in a CLOB column. Is there any way for me to filter by json value?

My Model

```
class Test(Base):
    __tablename__ = "test_data"
    id = Column(Integer, primary_key=True)
    j_json = Column(CLOB)

```

I tried something like that, but it's not working.
```
stmt = session.query(Test).filter(Test.j_json.Zuname == 'Blah').all()
stmt = select(Test).where(Test.j_json['Zuname'].as_string() == 'Blah')
```

Is it even possible to do in ORM or i have to go with row sql queries?

Thanks in advance

**Comment by Oynowan** [comment]
I found the solution for Oracle version 12 or later. 

`stmt = session.query(Test).filter(text("json_value(j_json, '$.Zuname') = 'Blah'")).all()`

The `json_value` function.

Still not sure if there is something simillar for a lower version, like 11.

**Reply by zzzeek** [reply]
you'd preferably use `func.json_value()`, see [functions](https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#working-with-sql-functions)

**Comment by Oynowan** [comment]
Thank you for your answer. Is there also a way to add constraint 'is json' to clob column? So i could also query my json data with dot notation from dbeaver. For now i can only with JSON_VALUE function, because oracle requires to specify constraint to clob as it 'is json'.

**Reply by CaselIT** [reply]
Can you provide a link to the oracle documentation regarding that option?
That's something that can be added, but I'm not familiar with it, so looking for the docs to look at what it would require on sqlalchemy side

**Reply by Oynowan** [reply]
Here is to oracle https://docs.oracle.com/database/121/ADXDB/json.htm#ADXDB6371.
I have found that oracledb has it too. https://python-oracledb.readthedocs.io/en/latest/user_guide/json_data_type.html

I implemented it on my own, i just wrote a custom type and creation is working, has to test it with orm selections and updates if it will work the same.


```
class OracleJson(types.UserDefinedType):
      cache_ok = True
      
      def __init__(self, col_name, constraint_name):
          self.col_name = col_name
          self.constraint_name = constraint_name
      
      def get_col_spec(self, **kw):
          return f"CLOB CONSTRAINT {self.constraint_name} CHECK ({self.col_name} IS JSON )"
      
      def bind_processor(self, dialect):
          def process(value):
              return value
          return process
  
      def result_processor(self, dialect, coltype):
          def process(value):
              return value
          return process

```

**Reply by CaselIT** [reply]
Thanks for the docs links.

It seems like this is more a constraint than an type option, so not sure if an `is_json=True` would be appropriate on the type as this could cause issues with alembic in autogenerate if it sees that the db has a constraint but there is none on the table definition.
What do you think @zzzeek ?

**Reply by zzzeek** [reply]
Oracle has a JSON datatype as of 21c, apparently.  We can support that directly.  

beyond that is the question..,how to make a type with a CHECK constraint?   You could try creating a SchemaType subclass.  here's a demo that works

```py
from sqlalchemy import CheckConstraint
from sqlalchemy import CLOB
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import text
from sqlalchemy import TypeDecorator
from sqlalchemy.orm import declarative_base
from sqlalchemy.types import CLOB
from sqlalchemy.types import SchemaType
from sqlalchemy.types import TypeDecorator


# inherit order needs to be TypeDecorator, SchemaType.
class OracleJSON(TypeDecorator, SchemaType):
    impl = CLOB

    def _set_table(self, column, table):
        e = CheckConstraint(
            text(f"{column.name} is json"),

            # not strictly necessary but helps if you are using
            # table.to_metadata
            _type_bound=True,
        )


        # or use table.append_constraint(e)
        e._set_parent(column)

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(OracleJSON)


e = create_engine("oracle://scott:tiger@oracle18c", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)
```

output

```sql
CREATE TABLE a (
	id INTEGER NOT NULL, 
	data CLOB CHECK (data is json), 
	PRIMARY KEY (id)
)
```

**Reply by CaselIT** [reply]
> Oracle has a JSON datatype as of 21c, apparently. We can support that directly.

that sounds like a plan. docs for reference https://docs.oracle.com/en/database/oracle/oracle-database/21/sqlrf/Data-Types.html#GUID-AB053D2C-2A40-478E-82E5-B9176C8776FD

We could add that type in oracle, and maybe your snippet in the oracle dialect docs

**Reply by Oynowan** [reply]
Unfortunately for a long time we had 11.2 Oracle version, for like last 5 years, and we just moved to version 19, so another move to 21c will take some time, thats why i can't use Oracles JSON type. Anyway, it would be good if you could support it in the future. For now custom type works. @zzzeek it didn't work for me, when autogenerated with alembic and ran upgrade, got error `AttributeError: 'OracleJson' object has no attribute 'get_col_spec'`, but thanks to you guys i have modified my solution and deleted constraint so one less variable to pass and it's working.

**Reply by CaselIT** [reply]
Well if using the constraint that can be added like any other, so it does not need a custom type, it's just clob, at least I think.

I don't know it on sqlalchemy side it would need some special handling to support rendering some of the json functionalities

**Reply by Oynowan** [reply]
Yea, i am having an issue now. When trying to query through, i am getting an error `NotImplementedError: Operator 'getitem' is not supported on this expression`. 

My custom type
```
class OracleJson(types.UserDefinedType):
    cache_ok = True
    
    def __init__(self, col_name):
        self.col_name = col_name
    
    def get_col_spec(self, **kw):
        return f"CLOB CHECK ({self.col_name} IS JSON )"
    
    def bind_processor(self, dialect):
        def process(value):
            return value
        return process

    def result_processor(self, dialect, coltype):
        def process(value):
            return value
        return process
```
Implementation into the model 
`json_data = Column(OracleJson('json_data'))`
Test query 
`res = session.query(ValidatedTable).filter(ValidatedTable.json_data['Name'] == 'TEST').all()`

Tried with this type too, same error.
```
class OracleJson(TypeDecorator, SchemaType):
    impl = CLOB

    def _set_table(self, column, table):
        e = CheckConstraint(
            text(f"{column.name} is json"),

            # not strictly necessary but helps if you are using
            # table.to_metadata
            _type_bound=True,
        )


        # or use table.append_constraint(e)
        e._set_parent(column)
        
    def get_col_spec(self, **kw):
        return f"CLOB CHECK ({self.col_name} IS JSON )"
```

**Reply by Oynowan** [reply]
It works when i go like that. `session.query(ValidatedTable).filter(func.json_value(ValidatedTable.json_data, text("'$.Name'")) == 'TEST').all()`
Am i doing something wrong way that i have to put single and double quotation marks `func.json_value(ValidatedTable.json_data, text("'$.Name'")`? If i will go just with `""` or `''`  i am getting an error `sqlalchemy.exc.DatabaseError: (oracledb.exceptions.DatabaseError) ORA-40454: path expression not a literal`

**Reply by zzzeek** [reply]
the `get_col_spec()` error only happens if you are using `UserDefinedType`.  You don't need UserDefinedType here because you have CLOB.

When I use this type, exactly:

```py
class OracleJSON(types.TypeDecorator, types.SchemaType):
    impl = types.CLOB

    def _set_table(self, column, table):
        e = CheckConstraint(
            text(f"{column.name} is json"),

            # not strictly necessary but helps if you are using
            # table.to_metadata
            _type_bound=True,
        )


        # or use table.append_constraint(e)
        e._set_parent(column)

```

running an autogenerate gives me a migration like:

```py
def upgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('test',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('data', env_py.OracleJSON(), nullable=True),
    sa.PrimaryKeyConstraint('id')
    )
    op.drop_table('user')
    op.drop_table('a')
    # ### end Alembic commands ###
```

noting that custom types in alembic have to have an import path of some kind.

Running that upgrade (after making sure OracleJSON is importable in the migration file) succeeds with SQL:

```sql
CREATE TABLE test (
	id INTEGER NOT NULL, 
	data CLOB CHECK (data is json), 
	PRIMARY KEY (id)
)


```


so from a DDL perspective, this custom type works on my end.    

Adding JSON operator support is a different matter.

**Reply by zzzeek** [reply]
for JSON operator support, figure out exactly what SQL that Oracle wants first, like use sql*plus or your console tool to get the SQL exactly right first.  then we can illustrate how SQLAlhcemy func can render it.

**Reply by Oynowan** [reply]
Hi, i am back after almost 3 months. Had a lot of work, but still same issue. 

So the oracle want something like this:
`SELECT td.json_data.test FROM test_data td;` 
this will retrieve data from column json_data from field test. Of course constraint is needed too, but this we have done already.

**Reply by Oynowan** [reply]
It's working now. I did something like this.
```
product = self.session.query(ProductTable).filter(text(f"json_value(JSON_DATA, '$.dwh_object_id') = {dwh_object_id}")).first()
```
It's working just fine, but would there be a solution for that string format? Instead of hard codding 'dwh_object_id', could i parse it as parameters? I am just not sure where to do it with 'text'. Couldn't find a solution. I know i could do it with execution, but what about filter?

**Comment by zzzeek** [comment]
OK so this looks like they are using MySQL's general JSON style, so below I've extracted all the bits from our MySQL implementation to create a full recipe.   This obviously requires a lot of knowledge of how things work to come up with, however, these are all public extension points and there's not actually much going on here that is unusual.   This will set up full JSON round trips if you use `JSON.with_variant(OracleJSON, "oracle"))`.

```py
import json
from typing import Any

from sqlalchemy import Any
from sqlalchemy import BinaryExpression
from sqlalchemy import BindParameter
from sqlalchemy import CLOB
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Dialect
from sqlalchemy import Identity
from sqlalchemy import Integer
from sqlalchemy import JSON
from sqlalchemy import select
from sqlalchemy import type_coerce
from sqlalchemy import TypeDecorator
from sqlalchemy.dialects import oracle
from sqlalchemy.engine.interfaces import Dialect
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.orm import Any
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.sql import operators
from sqlalchemy.sql.expression import BinaryExpression
from sqlalchemy.types import TypeEngine


Base = declarative_base()


class OracleJSON(TypeDecorator):
    """in order to send Oracle JSON types round trip, we need to coerce to
    string and back for DB round trips

    """

    impl = CLOB

    def process_bind_param(self, value: Any | None, dialect: Dialect) -> Any:
        if value is not None:
            value = json.dumps(value)
        return value

    def process_result_value(
        self, value: Any | None, dialect: Dialect
    ) -> Any | None:
        if value is not None:
            value = json.loads(value)
        return value


class Test(Base):
    __tablename__ = "test_data"
    id = Column(Integer, Identity(), primary_key=True)

    # using JSON() first establishes json-index operator support in Python.
    # can also add these to the OracleJSON type directly but that's more
    # intricate
    # then when data is being processed / compiled, OracleJSON is used
    j_json = Column(JSON().with_variant(OracleJSON(), "oracle"))


class _JSONExprMixin(TypeEngine):
    """this mixin is copied from the MySQL dialect.

    it applies the "$." or similar syntax to the index values.

    """

    def _format_value(self, value):
        raise NotImplementedError()

    def bind_processor(self, dialect):
        super_proc = super().bind_processor(dialect)

        def process(value):
            value = self._format_value(value)
            if super_proc:
                value = super_proc(value)
            return value

        return process

    def literal_processor(self, dialect):
        super_proc = super().literal_processor(dialect)

        def process(value):
            value = self._format_value(value)
            if super_proc:
                value = super_proc(value)
            return value

        return process


class OracleJSONIndexType(_JSONExprMixin, JSON.JSONIndexType):
    def _format_value(self, value):
        if isinstance(value, int):
            value = "$[%s]" % value
        else:
            value = '$."%s"' % value
        return value


class OracleJSONPathType(_JSONExprMixin, JSON.JSONPathType):
    """if Oracle does JSON paths the same way MySQL does also,
    you can use this too"""

    def _format_value(self, value):
        return "$%s" % (
            "".join(
                [
                    "[%s]" % elem if isinstance(elem, int) else '."%s"' % elem
                    for elem in value
                ]
            )
        )


@compiles(BinaryExpression, "oracle")
def visit_binary(binary, compiler, **kw):
    """intercept the JSON 'getitem' operator and deliver special processing"""
    if binary.operator is operators.json_getitem_op:
        # this code assumes a string or integer literal index value, so
        # assert that here
        assert isinstance(binary.right, BindParameter)
        assert isinstance(binary.right.type, JSON.JSONIndexType)

        return "JSON_VALUE(%s, %s)" % (
            compiler.process(binary.left, **kw),
            compiler.process(
                type_coerce(
                    # oracle requires a "literal value" here, so transform
                    # the parameter to render as a "literal"
                    binary.right.render_literal_execute(),

                    # in order to get our _JSONExprMixin._format_value
                    # to take effect, coerce the type of the expression
                    # to be our OracleJSONIndexType, instead of the plain
                    # JSON.JSONIndexType
                    OracleJSONIndexType
                ),
                **kw,
            ),
        )

    # compile other expressions normally
    return compiler.visit_binary(binary, **kw)


stmt = select(Test).filter(Test.j_json["Zuname"] == "Blah")

# do a dry run
print(
    stmt.compile(
        dialect=oracle.dialect(), compile_kwargs={"literal_binds": True}
    )
)


e = create_engine("oracle://scott:tiger@oracle18c", echo=True)

Base.metadata.drop_all(e)
Base.metadata.create_all(e)

with Session(e) as sess:
    # INSERT some json
    t1 = Test(j_json={"Zuname": "Blah"})
    t2 = Test(j_json={"Zuname": "not blah"})
    sess.add_all([t1, t2])
    sess.commit()

    # SELECT some json with the index thing
    t1_returned = sess.scalars(stmt).one()
    assert t1_returned is t1

    # assert the json was coerced too
    assert t1_returned.j_json == {"Zuname": "Blah"}

```

**Reply by CaselIT** [reply]
Is this something that can be integrated in SQLAlchemy?

**Reply by zzzeek** [reply]
of course, it means we'd implement Json for Oracle mostly like we do for MySQL / SQLite.

though the naming thing means we'd have to call it something different like oracle.JsonCLOB.

**Reply by zzzeek** [reply]
oh, this just in, SQL Server is breaking the rule already with JSON -> NVARCHAR.

oh well   .  JSON should be renamed "Json".    in the meantime the Oracle version can be implemented directly

**Reply by Oynowan** [reply]
Thank you for that, looks good. I just have to adjust alembic for that too, because i am using autogenerator for migrations and alembic is using straight only OracleJson
`sa.Column('j_json', custom_type.OracleJson(), nullable=True)`
instead of 
`j_json = Column(JSON().with_variant(OracleJSON(), "oracle"))`.
It's a process which has to be fully automated. Probably i will implement those solutions in next iteration of code. Is there possibility that you would implement that JSON solution for Oracle into standard sqlalchemy in near future? We are planning to move to postgres, but we are big company and it will take some time for sure before it happen.
